cms/djangoapps/contentstore/views/preview.py

```diff
@@ -216,7 +216,6 @@ def _preview_module_system(request, descriptor, field_data):
             "settings": SettingsService(),
             "user": DjangoXBlockUserService(
                 request.user,
-                anonymous_user_id='student',
```

Comment:
@schenedx This is not enough to fix the issue.. you'll need to actually compute the `anonymous_user_id` for the current user, and pass it into the UserService as is [done by the LMS ModuleSystem](https://github.com/openedx/edx-platform/pull/29190/files#diff-3b0c1e091e61826c4448bc5924e2911dedfd829754b2c6087e7067dfbdeed895R541-R550).

Alternatively, you can move the logic for computing this value into a new function on the UserService (it will need access to the module `descriptor` and the `course_id`) to avoid duplicating code.

Reply:
Nudge @schenedx :) Do you have an estimate for when you'll be able to address my comments?
Just planning my week, and want to make sure I allow time for a 2nd review. Thanks!

Reply:
No worries @schenedx :)

> Also, this is on CMS. The LMS side has the correct value returned.

Yes. What I'm saying above is the CMS needs to return the same value as the LMS, and this requires some code changes.

> I thought the DjangoXBlockUserService instance already have the [get_anynomous_user_id](https://github.com/openedx/edx-platform/blob/db32ff2cdf678fa8edd12c9da76a76eef0478614/common/djangoapps/xblock_django/user_service.py#L66-L87) that returns the calculated value, no? ([sic: corrected source link])

Yes it does have that method, but there's some issues with using it directly:

* performance: that method hits the database to get the User object for the given username, so we'd likely want to cache it
* There's some subtleties of the [LMS code](https://github.com/openedx/edx-platform/blob/db32ff2cdf678fa8edd12c9da76a76eef0478614/common/djangoapps/xblock_django/user_service.py#L66-L87) -- some blocks expect a per-student anonymous ID, while some expect a per-course anonymous ID. The `DjangoXBlockUserService ` currently has no knowledge of `course_id`, or of the block's `requires_per_student_anonymous_id` flag.

All of these issues can be solved, but they require code changes to implement them.

> I have no developer experience in this area, so this is difficult for me to setup.

Ah ok.. I misunderstood. Do you have time and interest in making this change? If not, then we should find someone who does. OpenCraft could do it (for a fee), if there's no one in the community who wants to do it?

> How do we test the anynomous_user_id with randomized content? Do I just need to set it up on my local environment? 

Yep, if you want to implement this change, you'll need a [devstack](https://github.com/openedx/devstack). You can test it manually, but we'd also want some automated tests to protect the change, and ensure that future code changes don't break it.

> Also, I think all the side effect of this change @pomegranited have listed is worth a heads up for 2U internal. 

Yep, that's worth doing, and can be done as part of the OSPR too. It's always easier to see the effects of a change when the code is in place so it can be tested.